### ISSUE TITLE
Jetty maven plugin reload is triggered each time the `scanIntervalSeconds` pass

### ISSUE THREAD

**Issue by latuszek** [eclipse/jetty.project#2266]
Hi,

I'm running jetty via `mvn clean jetty:run` and each time the `scanIntervalSeconds` pass, jetty trigger hot reload - with no changes made by me in the meantime. This is happening in plugin version 9.4.7 and 9.4.8, therefore I stuck at 9.4.6 for now. I would gladly provide some more insights if needed, just point me to the right direction.

My plugin config:
```xml
                    <plugin>
                        <groupId>org.eclipse.jetty</groupId>
                        <artifactId>jetty-maven-plugin</artifactId>
                        <configuration>
                            <scanIntervalSeconds>10</scanIntervalSeconds>
                            <systemProperties>
                                <systemProperty>
                                    <name>SERVICE_ENVIRONMENT</name>
                                    <value>dev</value>
                                </systemProperty>
                            </systemProperties>
                        </configuration>
                        <dependencies>
                            <!-- colored logs for logback -->
                            <dependency>
                                <groupId>org.fusesource.jansi</groupId>
                                <artifactId>jansi</artifactId>
                                <version>1.17</version>
                            </dependency>
                        </dependencies>
                    </plugin>
```

I'm facing this issue since version 9.4.7 of jetty-maven-plugin (I believe the cause may be someware in this commit: 20ca6277a8e49fa53932908f845b740182e88229).

**Comment by olamy** [comment]
which os are you using?

**Comment by seiyapl** [comment]
Windows 10, but I'm running maven from git bash usually.

**Comment by olamy** [comment]
@seiyapl is it a multi module project? By chance is it a project you can share?

**Comment by olamy** [comment]
I just tried with this simple project: https://github.com/olamy/helloworld-servlet
I cannot reproduce on osx neither linux/ubuntu.
I will try with a multi module project

**Comment by seiyapl** [comment]
@olamy I cannot share my project. But I can share logs or pom.xml if it may help?
I have only one pom.xml in my project.

**Comment by olamy** [comment]
please share your pom and logs?
can you try my sample project as well?
well I have to find a windows laptop :-)

**Comment by seiyapl** [comment]
I've tried Your project and it was not restarting like mine.

Here are logs and pom from my project:
https://drive.google.com/open?id=1wqktSids6VA4W5TH2Gui4vqDkpWXsSh7

**Comment by seiyapl** [comment]
@olamy any thoughts?

**Comment by olamy** [comment]
@seiyapl which OS are you using?

**Comment by seiyapl** [comment]
@olamy 

> Windows 10, but I'm running maven from git bash usually.

**Comment by seiyapl** [comment]
@janbartel I will do this today. BTW have You seen attached logs in my previous post?

**Comment by janbartel** [comment]
@seiyapl yes, I've had a look.  Nothing to indicate why a restart is happening. You do have some (unrelated) issues about your classpath though - see all the messages warning about scanning same class from different sources: you have javax.mail api classes included in the javax.mail impl jar so they're duplicated, and you must have put the jstl jars as dependencies, even though it is supplied by the container.

You can also try running with the latest version of jetty, which is jetty-9.4.10.RC1.

**Comment by seiyapl** [comment]
@janbartel Unfortunately it was false-positive.
Logs from `mvn -X jetty:run`:  https://drive.google.com/file/d/1O6I61sb6g8hLMSXwIQF_tWua9UWkBobC/view?usp=sharing

**Comment by latuszek** [comment]
Issue still occurs in 9.4.11.

**Comment by latuszek** [comment]
I will try to extract something that I can share from my project, but it will take me some time as we are on a tight schedule.

**Comment by NikitaKriukov** [comment]
@olamy I faced similar problem and noticed a curious details, which can be important. 
In my case Jetty is reloaded with specific conditions: 
1. The same maven project is opened in IntelliJ. 
2. IntelliJ window should be maximized! 
3. mvn jetty:run should be run at an external terminal window
4. It's necessary to switch to IntelliJ. Be aware, that IntelliJ window should be maximised!
5. Switch back to the terminal
6. Jetty is reloaded. 

Jetty also reloaded when the IntelliJ is started or stopped. It looks like they disturb each others someway.

**Comment by NikitaKriukov** [comment]
@olamy I found that IntelliJ saves project when it lost focus. And Jetty plugin by default takes changes. It can be fixed with adding <configuration><reload>manual</reload></configuration>

**Comment by janbartel** [comment]
@NikitaKriukov that makes perfect sense: the jetty maven plugin is designed to restart the webapp when the IDE changes it.

**Comment by latuszek** [comment]
@janbartel but should it reload the project if the files doesn't change at all? In my case I'm not changing anything. I'm using IntelliJ also, but I'm not making any changes to files between reloads. 
I'm still on version 9.4.6 of the plugin as the 9.4.14 still has the issue.
Sorry that I didn't extract my project, but we have so tight schedule that it is not possible.
Have You looked to the logs I've attached on 9 May?

**Comment by gregw** [comment]
@latuszek many users consider it a feature that you can just touch a file to trigger a redeploy, so I don't think we should go to the effort of scanning/checksumming the files.

**Comment by janbartel** [comment]
OK, let me walk that back a bit. As @gregw  has pointed out to me, it has always been the case that we  can 'touch' a file and have the plugin redeploy a file. So that behaviour should not have changed.

So the only change there was the replacement of Scanner with PathWatcher, but both should give the same result.

**Comment by latuszek** [comment]
@gregw I agree. But the issue I have, did not occurred with IntelliJ version being updated, but the jetty-maven-plugin. And in version 9.4.6 it still works correctly. Since 9.4.7 the application is being reloaded for unknown (for me) reason - meaning I'm not changing my files. I might switch the focus from the editor tab to any other. I'm not sure what IntelliJ is doing under the hood, but it did not trigger the reload until 9.4.7. For me it's a blocker as the spring context reload takes around 15 seconds each time I do anything in the IntelliJ.

@janbartel Thanks for the research. Unfortunately I will not be able to provide a clean reproducible environment for the tests but I will happily test any fix for the issue on my environment.

**Comment by latuszek** [comment]
@gregw Am I correct that Your [change](https://github.com/eclipse/jetty.project/commits/80fb788d0c152c0628e1292f9d571baee720e336/jetty-util/src/main/java/org/eclipse/jetty/util/PathWatcher.java) might alter this behavior? Would You be able to help?

**Comment by gregw** [comment]
@latuszek It is my understanding that we've always been able to just touch a file to redeploy, so my change should not have changed that.... unless it had somehow been broken before.
We can research that change in behaviour, but more importantly I believe that we want that as a feature.  Being able to touch to redeploy is a wanted feature.

Can you give us a full listing of your webapp directory before and after intellij loses focus so we can see what files change.   Perhaps we can give you some configuration to ignore some files and only redeploy on other ones, but ultimately I think this is something intellij should fix.  They should not be modifying files that have not changed simply before of focus!

**Comment by latuszek** [comment]
@gregw I'm not sure if this is going in a good direction. I can send you those listings, however the facts are that those redeploys:
1. does not occur in the jetty-maven-plugin version `9.4.6.v20170531` and below - I'm still using this version (and I was updating IntelliJ many times since the 9.4.6 release).
2. does occur in all versions starting from `9.4.7.v20170914` up to the `9.4.14.v20181114`.
3. does not depend on IntelliJ version.

Based on this observations I'm pretty sure that there must be some change in the jetty-maven-plugin code (between 9.4.6 and 9.4.7) that triggers this behavior.

> Being able to touch to redeploy is a wanted feature.
If this feature existed before v9.4.7 I think it's not a problem.

**Comment by latuszek** [comment]
@gregw, @joakime did you have a time to look at this? any luck? 

@joakime It may not be an obvious change, but it's obvious (at least to me) that there must be some change that cause this issue.

**Comment by joakime** [comment]
Not reproducible on OSX / Java 11.0.2 / Jetty 9.4.15.

**Comment by latuszek** [comment]
Still occurs on Windows 10 / Java 1.8.0_201 / jetty-maven-plugin 9.4.15.v20190215 / IntelliJ 2018.3.4

**Comment by janbartel** [comment]
@latuszek  if you change the scan interval to something bigger, like 1 minute does it still occur?

**Comment by latuszek** [comment]
I've tested it a bit and the scan interval seem to not matter at all. I've set it to 1 second and for 600 seconds - the result is the same.

I'm curious what is jetty checking to determine if the app should be reloaded or not? I'm asking because maybe jetty is OK but IntelliJ does something wrong - e.g. maybe it's marking files as modified even if nothing changes.

**Comment by joakime** [comment]
@latuszek as an experiment, compile this from oracle and run it against your project while you are editing in intellij.

https://docs.oracle.com/javase/tutorial/essential/io/examples/WatchDir.java

That will tell you what kinds of events are seen by the FileSystem and Java's `java.nio.files.WatchService` (which Jetty uses).
It might be that Jetty is responding correctly to your level of filesystem changes, and instead of configuring Jetty, it would make more sense to configure IntelliJ to be less aggressive in its updates.

Oh, and one more thing I remembered, if your code is on a non-native filesystem (such as NFS, Windows Share, Virtual FileSystems, Rational ClearCase, USB, Thunderbolt, etc) then you'll experience a completely different class of behaviors then what is normally expected with filesystem updates (ClearCase and NFS being the biggest culprits in both extreme conditions: over-notification, dropped-notification).

**Comment by latuszek** [comment]
@joakime Here's what this class logged in the moment of the reload:
```
ENTRY_CREATE: C:\workspace\code\cargo\.idea\workspace.xml___jb_tmp___
ENTRY_MODIFY: C:\workspace\code\cargo\.idea
ENTRY_MODIFY: C:\workspace\code\cargo\.idea\workspace.xml___jb_tmp___
ENTRY_DELETE: C:\workspace\code\cargo\.idea\workspace.xml
ENTRY_CREATE: C:\workspace\code\cargo\.idea\workspace.xml___jb_old___
ENTRY_DELETE: C:\workspace\code\cargo\.idea\workspace.xml___jb_tmp___
ENTRY_CREATE: C:\workspace\code\cargo\.idea\workspace.xml
ENTRY_MODIFY: C:\workspace\code\cargo\.idea\workspace.xml
ENTRY_DELETE: C:\workspace\code\cargo\.idea\workspace.xml___jb_old___
ENTRY_MODIFY: C:\workspace\code\cargo\.idea
ENTRY_MODIFY: C:\workspace\code\cargo\target\tmp
ENTRY_DELETE: C:\workspace\code\cargo\target\tmp\jsp
ENTRY_MODIFY: C:\workspace\code\cargo\target\tmp
ENTRY_DELETE: C:\workspace\code\cargo\target\tmp
ENTRY_MODIFY: C:\workspace\code\cargo\target
ENTRY_CREATE: C:\workspace\code\cargo\target\tmp
register: C:\workspace\code\cargo\target\tmp
ENTRY_MODIFY: C:\workspace\code\cargo\target
```

**Comment by DevVito** [comment]
Hi All,

I had a chance to take a look at @latuszek's project and noticed an interesting difference between the two versions of jetty plugin, which may be a useful hint. Namely, they differ in the scope of directories scanned for changes. I tested it having IDE turned off and running maven from the command line on Windows 10. To trigger changes detection I created a new, empty text file in a particular directory. The results show if directory has been scanned or not: 

Dir | v 9.4.6.v20180619 | v 9.4.7.v20180619
------------ | ------------- | --------------
project | no | no
project/.idea|no|**yes** (excluding subdirectories)
project/.mvn|no|**yes** (e.s.)
project/src|no|**yes** (e.s.)
project/src/main|no|no
project/src/main/webapp|no|no
project/src/test|no|no
project/target|no|**yes**
project/target/classes|yes (i.s.)|yes (including subdirectories)
project/target/generated-sources|no|no
project/target/test-classes|no|no

I suspect, the behavior which both @latuszek and @NikitaKriukov faced using IntelliJ was directly caused by the scanning of `project/.idea` directory where `workspace.xml` file changes every time IntelliJ performs the "save" action, which - by default - is triggered automatically every time IntelliJ window gets or loses focus.

Is it possible to exclude the `.idea` folder from scanned scope by configuration? I think that could serve as a workaround without code changes...

**Comment by janbartel** [comment]
@DevVito thanks for that info, will look into it and report back on this issue.

**Comment by dulakm** [comment]
@janbartel any update on this? 

The issue still persists for `9.4.20.v20190813`. Thank you!

**Comment by janbartel** [comment]
@dulakm and @DevVito sorry for the delay - this issue got away from me.

Seems bizarre behaviour: on linux, I can't reproduce the list of changes that @DevVito posted on 26th April using the plugin. So I'm thinking this is some different behaviour in windows?  I'll add another PathWatcher test today just to double-check and get it run on windows.

**Comment by janbartel** [comment]
I've found the problem: seems to be in the PathWatcher class behaviour. I've raised issue #4218 

I'll see if I can create a workaround for the plugin.

**Comment by janbartel** [comment]
Fix committed for jetty-49.4.23.
